File: my_cpp_pkg/src/battery.cpp
#include "rclcpp/rclcpp.hpp"
#include "my_robot_interfaces/srv/set_led.hpp"

class BatteryNode : public rclcpp::Node // MODIFY NAME
{
public:
    BatteryNode() : Node("battery") // MODIFY NAME
    {   
        low_on_battery_ = false;
        last_change_time_ = this->get_clock()->now().seconds();
        timer_ = this->create_wall_timer(std::chrono::milliseconds(100),
                                         std::bind(&BatteryNode::callback_set_led, this));
        RCLCPP_INFO(this->get_logger(), "Battery node has been launched.");
    }

private:
    void call_set_led(int led_number, bool led_state)
    {
        auto client_ = this->create_client<my_robot_interfaces::srv::SetLed>("set_led");

        while (!client_->wait_for_service(std::chrono::seconds(1)))
        {
            RCLCPP_WARN(this->get_logger(), "Waiting for the server...");
        }
        auto request_ = std::make_shared<my_robot_interfaces::srv::SetLed::Request>();
        request_->led_number = led_number;
        request_->led_state = led_state;

        auto future = client_->async_send_request(request_);
        try
        {
            auto response = future.get();
        }
        catch (const std::exception &e)
        {
            RCLCPP_ERROR(this->get_logger(), "Service call failed.");
        }
    }

    void callback_set_led()
    {   double time_now_ = this->get_clock()->now().seconds();
        if (low_on_battery_ && (time_now_-last_change_time_) > 6.0){
            low_on_battery_ = false;
            RCLCPP_INFO(this->get_logger(), "Battery charged!");
            threads_.push_back(std::thread(std::bind(&BatteryNode::call_set_led, this, 3, false)));
            last_change_time_ = time_now_;
        }else if (!low_on_battery_ && (time_now_-last_change_time_) > 4){
            low_on_battery_ = true;
            RCLCPP_INFO(this->get_logger(), "Battery is empty! Charging...");
            threads_.push_back(std::thread(std::bind(&BatteryNode::call_set_led, this, 3, true)));
            last_change_time_ = time_now_;
        }
        
    }

    std::vector<std::thread> threads_;
    rclcpp::TimerBase::SharedPtr timer_;
    double last_change_time_;
    bool low_on_battery_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<BatteryNode>(); // MODIFY NAME
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: my_py_pkg/my_py_pkg/battery.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from my_robot_interfaces.srv import SetLed
from functools import partial


class BatteryNode(Node): # MODIFY NAME
    def __init__(self):
        super().__init__("battery") # MODIFY NAME
        self.last_change_time = self.get_current_time()
        self.low_on_battery = False
        self.get_logger().info("Battery node has been launched.")
        self.create_timer(0.1, self.callback_change_battery_charge)

    def get_current_time(self):
        secs, nanosecs = self.get_clock().now().seconds_nanoseconds()
        return secs + (nanosecs/1000000000.0)

    def callback_change_battery_charge(self):
        time_now = self.get_current_time()
        if (not(self.low_on_battery) and (time_now - self.last_change_time) > 4.0):
            self.low_on_battery = True
            self.get_logger().info("Battery is empty! Charging...")
            self.call_set_led(True, 3)
            self.last_change_time = time_now
        elif (self.low_on_battery and (time_now - self.last_change_time) > 6.0):
            self.low_on_battery = False
            self.get_logger().info("Battery charged!")
            self.call_set_led(False, 3)
            self.last_change_time = time_now
        

    def call_set_led(self, battery_low, led_number):
        client_ = self.create_client(SetLed, "set_led")
        while not client_.wait_for_service(1):
            self.get_logger().warn("Waiting for Server Set Led...")
        request = SetLed.Request()
        request.led_number = led_number
        request.led_state = battery_low

        future = client_.call_async(request)
        future.add_done_callback(
            partial(self.callback_set_led, led_number=led_number, led_state = battery_low))

    def callback_set_led(self, future, led_number, led_state):
        try:
            response = future.result()
            if (response):
                self.get_logger().info("Led has been updated.")
            else:
                self.get_logger().info("Led has NOT been updated.")
        except Exception as e:
            self.get_logger().error("Service call failed %r"%(e, ))

def main(args=None):
    rclpy.init(args=args)
    node = BatteryNode() # MODIFY NAME
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()
